Design/y_e_data_storage_modes.md

```diff
@@ -0,0 +1,151 @@
+
+## Uncertainties
+
+### Status
+
+- Typically, `E` stores the standard deviation.
+- For computations such as adding data, the variance is required, thus `E` is squared and converted back to standard deviation by taking the `sqrt` at a later points.
+  Often this is done in-place, i.e., `E` temporarily stores the variance.
+- In a few cases `E` is kept as a variance until just before the end of an algorithm for performance reasons.
+  For example, `DiffractionFocussing2` needs to summing many spectra.
+  To avoid repeated evaluation of the expensive `sqrt` function for the target spectrum, all spectra are added without converting back to standard deviation after every addition. `DiffractionFocusing2.cpp` calls `Kernel::VectorHelper::rebinHistogram()` where calculations between intput and output E `vector` are performed in terms of variance. This also happens in `Kernel::VectorHelper::rebin()` 
+
+### Discussion
+
+For `E` data, the modification of the storage mode is typically very local and a fix should be rather easy.
+
+- `Histogram` currently provides functions for obtaining variances as well as standard deviations, e.g., `countVariances()` and `countStandardDeviations()`.
+  If the internal `E` vector were converted to variances by external means (as it is done currently), these access methods would do the wrong thing.
+
+- `HistogramData` provides means for converting between standard deviations and variances, in particular types that do the conversion automatically.
+  These should be used by client code.
+
+- If summing spectra were done using `Histogram::operator+()`, propagating the uncertainties would involve a `sqrt` computation for every `+` operation.
+  In the example of `DiffractionFocussing2` this could yield a significant overhead, and it is not obvious how to solve this:
+
+  - The algorithm could do the summing itself, summing `Y` and variances separately.
+    This would cause no performance loss, and the code would be reasonably clean, e.g.,
+
+    ```cpp
+    for (size_t i=0; i<inputWS.getNumberHistograms(); ++i) {
+      counts += inputWS.counts(i);
+      variances += inputWS.countVariances(i);
+    }
+    ```
+
+  - We could provide optimized means for summing many spectra as part of the `HistogramData` module.
+    For example,
+
+    ```cpp
+    namespace HistogramData {
+    Histogram sumHistograms(const std::vector<Histogram> &histograms);
+    }
+    ```
+
+    This would be a clean solution without performance loss.
+    Unfortunately, there will be an implicit performance and ressource issue if the input is an `EventWorkspace`!
+    Obtaining a `Histogram` from the input workspace will create the `Y` and `E` data on the fly (this needs to be done in any solution).
+    However, to create the vector of histograms as input for `sumHistograms()` we need to **store all histograms at the same time**.
+    This will increase memory consumption and reduce cache locality of the computation.
+
```

Comment:
@OwenArnold I think I favor providing only `operator+` and not `sumHistograms()`. In the special cases where performance matters (like `DiffractionFocussing`) it should not be much trouble to hand-write the addition (still simpler than it is now, since we will have `CountVariances::operator+=`).